### PR TITLE
COMP: true/false completion for bool values

### LINKED ISSUE #4443: No true/false autocomplete for boolean values
## Environment

* **IntelliJ Rust plugin version:** 0.2.106.2134-193
* **Rust toolchain version:** 1.37.0 (eae3437df 2019-08-13) x86_64-apple-darwin
* **IDE name and version:** CLion 2019.3 EAP (CL-193.3519.24)
* **Operating system:** macOS 10.14.6

## Problem description

When the variable has the `bool` type, I would expect `true` and `false` to be the first autocomplete suggestions.

Instead, no boolean values are proposed at all:
Screenshot 2019-09-25 at 16 45 17
Screenshot 2019-09-25 at 16 45 26
Screenshot 2019-09-25 at 16 45 30
Screenshot 2019-09-25 at 16 45 37

### PR DESCRIPTION
**Pull request by ozkriff**
Fixes #4443 (_"No true/false autocomplete for boolean values"_)

### CODE REVIEW

File: src/test/kotlin/org/rust/lang/core/completion/RsCompletionTest.kt [hunk 1]
```diff
@@ -1199,4 +1199,42 @@ class RsCompletionTest : RsCompletionTestBase() {
...
+        fn main() { let x: u8 = fal/*caret*/; }
+    """)
+
+    fun `test complete true in condition`() = doSingleCompletion("""
+        fn main() { if tr/*caret*/ }
+    """, """
+        fn main() { if true/*caret*/ }
+    """)
+
+    fun `test complete false when assigning to bool variable`() = doSingleCompletion("""
```
Comment (Kobzol) [inline src/test/kotlin/org/rust/lang/core/completion/RsCompletionTest.kt:1229]: The test name uses false, but the text completes true :)

Comment (ozkriff) [reply src/test/kotlin/org/rust/lang/core/completion/RsCompletionTest.kt:1229]: Fixed!

File: src/main/kotlin/org/rust/lang/core/completion/RsCommonCompletionProvider.kt [hunk 2]
```diff
@@ -227,20 +228,6 @@ object RsCommonCompletionProvider : RsCompletionProvider() {
     override val elementPattern: ElementPattern<PsiElement>
         get() = PlatformPatterns.psiElement().withParent(psiElement<RsReferenceElement>())
 
-    private val simplePathPattern: ElementPattern<PsiElement>
-        get() {
-            val simplePath = psiElement<RsPath>()
-                .with(object : PatternCondition<RsPath>("SimplePath") {
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/completion/RsCommonCompletionProvider.kt:233]: looks like `PatternCondition` is unused import in this file now

File: src/main/kotlin/org/rust/lang/core/completion/RsBoolCompletionProvider.kt [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+import org.rust.lang.core.types.expectedType
+import org.rust.lang.core.types.ty.TyBool
+
+object RsBoolCompletionProvider : RsCompletionProvider() {
+
+    override val elementPattern: ElementPattern<PsiElement> get() = RsPsiPattern.simplePathPattern
+
+    override fun addCompletions(parameters: CompletionParameters, context: ProcessingContext, result: CompletionResultSet) {
+        if (parameters.position.ancestorOrSelf<RsPathExpr>()?.expectedType == TyBool) {
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/completion/RsBoolCompletionProvider.kt:25]:
```suggestion
        if (parameters.position.safeGetOriginalOrSelf().ancestorOrSelf<RsPathExpr>()?.expectedType == TyBool) {
```

`parameters.position` returns psi element in copy of original PSI file. `safeGetOriginalOrSelf()` returns element from original file if possible. It allows us to reuse existing type inference/name resolution caches if possible

File: src/test/kotlin/org/rust/lang/core/completion/RsCompletionTest.kt [hunk 4]
```diff
@@ -1199,4 +1199,42 @@ class RsCompletionTest : RsCompletionTestBase() {
     """, """
         fn foo(f: &FnOnce(/*caret*/)) {}
     """)
+
+    fun `test complete true in variable decl`() = doSingleCompletion("""
```
Comment (Undin) [inline src/test/kotlin/org/rust/lang/core/completion/RsCompletionTest.kt:1203]: I'd personally prefer to extract these tests into separate class
Just to make test search more convenient and don't create too large files

File: src/main/kotlin/org/rust/lang/core/completion/RsBoolCompletionProvider.kt [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+            result.addElement(LookupElementBuilder.create("false").bold().withPriority(KEYWORD_PRIORITY))
+            result.addElement(LookupElementBuilder.create("true").bold().withPriority(KEYWORD_PRIORITY))
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/completion/RsBoolCompletionProvider.kt:27]: Maybe it's better to wrap it into loop?
```kotlin
for (value in listOf("true", "false") {
    result.addElement(LookupElementBuilder.create(value).bold().withPriority(KEYWORD_PRIORITY))
}
```

Just not to duplicate code and simplify future changes

File: src/main/kotlin/org/rust/lang/core/completion/RsCommonCompletionProvider.kt (hunk 2 above)
Comment (ozkriff) [reply src/main/kotlin/org/rust/lang/core/completion/RsCommonCompletionProvider.kt:233]: Fixed

File: src/main/kotlin/org/rust/lang/core/completion/RsBoolCompletionProvider.kt (hunk 3 above)
Comment (ozkriff) [reply src/main/kotlin/org/rust/lang/core/completion/RsBoolCompletionProvider.kt]: Done!

File: src/main/kotlin/org/rust/lang/core/completion/RsBoolCompletionProvider.kt (hunk 5 above)
Comment (ozkriff) [reply src/main/kotlin/org/rust/lang/core/completion/RsBoolCompletionProvider.kt]: Done!

File: src/test/kotlin/org/rust/lang/core/completion/RsCompletionTest.kt (hunk 4 above)
Comment (ozkriff) [reply src/test/kotlin/org/rust/lang/core/completion/RsCompletionTest.kt]: Extracted

**Comment by ozkriff** [comment]
Should I rebase/squash the commits into one?

**Comment by Undin** [comment]
@ozkriff I suppose there are two meaningful changes: new completion contributor itself and new contributor in contributors.txt. So I'd personally make two commits. But if you want to squash them into a single one, it's also ok

**Comment by Undin** [comment]
bors retry

**Comment by ozkriff** [comment]
Yay!

**Pull request merged**